// json-tokenize/dist.js

var tokenTypes = [
  {
    regexp: /^\s+/,
    create: function (value, position) { return ({
      type: 'whitespace',
      position: position,
      raw: value,
      value: value
    }); }
  },

  {
    regexp: /^"(?:[^"\\]|\\.)*"/,
    create: function (value, position) { return ({
      type: 'string',
      position: position,
      raw: value,
      value: value
        .slice(1, -1)
        .replace(/\\"/g, '"')
    }); }
  },

  {
    regexp: /^(true|false|null)/,
    create: function (value, position) { return ({
      type: 'literal',
      position: position,
      raw: value,
      value: value === 'null'
        ? null
        : value === 'true'
    }); }
  },

  {
    regexp: /^(-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?)/,
    create: function (value, position) { return ({
      type: 'number',
      position: position,
      raw: value,
      value: +value
    }); }
  },

  {
    regexp: /^({|}|\[|]|:|,)/,
    create: function (value, position) { return ({
      type: 'punctuation',
      position: position,
      raw: value,
      value: value
    }); }
  }
]

var tokenize = function (
  json,
  tokens,
  position
) {
  if ( tokens === void 0 ) tokens = [];
  if ( position === void 0 ) position = { lineno: 1, column: 1 };

  var char = json[0]

  if (!char) {
    return tokens
  }

  var ref = tokenTypes.reduce(function (acc, type) {
    if (acc) { return acc }
    var str = match(type.regexp)
    if (!str) { return acc }

    return [
      type.create,
      str
    ]
  }, null);
  var createToken = ref[0];
  var str = ref[1];

  var token = createToken(
    str,
    str.length === 1
      ? position
      : { start: position, end: updateColumn(str.length - 1) }
  )

  var lines = str.match(/^\n+/g)

  if (lines) {
    return tokenize(
      advance(lines),
      tokens,
      { lineno: position.lineno + lines.length, column: 1 }
    )
  }

  return next(token)

  function updateColumn (column) {
    return {
      lineno: position.lineno,
      column: position.column + column
    }
  }

  function next (token, nextPosition) {
    return tokenize(
      advance(token.raw),
      tokens.concat([token]),
      nextPosition || updateColumn(token.raw.length)
    )
  }

  function match (re) {
    var m = re.exec(json)
    if (!m) { return }
    var str = m[0]
    return str
  }

  function advance (str) {
    return json.slice(str.length)
  }
}

module.exports = tokenize
